dist/aws/handler.py

```diff
@@ -1,144 +1,215 @@
-        bucket = event["Records"][0]["s3"]["bucket"]["name"]
+        return pipeline
 
-        file_key = urllib.unquote_plus(event["Records"][0]["s3"]["object"]["key"])
-        _json = ""
+    elif trigger == "S3":
+        print("Triggered from S3")
+
+        # get boto s3 client
+        s3 = boto3.client("s3")
+
+        # We always have to look to the bucket that triggered the event for the input
+        triggered_bucket = event["Records"][0]["s3"]["bucket"]["name"]
 
         # get the pipeline object and read
-        pipeline_obj = s3.get_object(Bucket=bucket, Key=file_key)
-        pipeline = json.loads(pipeline_obj["Body"].read().decode("utf-8"))
+        file_key = urllib.unquote_plus(event["Records"][0]["s3"]["object"]["key"])
+        pipline_obj = s3.get_object(Bucket=triggered_bucket, Key=file_key)
+        pipeline = json.loads(pipline_obj["Body"].read().decode("utf-8"))
 
-        # We can return if there is a valid cached output to save compute time.
-        settings_trigger = pipeline["settings"]
+        # provide consistent pipeline defaults
+        pipeline = default_pipeline(pipeline)
 
-        # check for cached output
-        cached = False
-        output_filename = file_key.replace(".json", "." + pipeline["output"]["format"]).replace(
-            settings_trigger["FUNCTION_S3_INPUT"], settings_trigger["FUNCTION_S3_OUTPUT"]
+        # create output filename
+        pipeline["output"]["filename"] = file_key.replace(".json", "." + pipeline["output"]["format"]).replace(
+            pipeline["settings"]["FUNCTION_S3_INPUT"], pipeline["settings"]["FUNCTION_S3_OUTPUT"]
         )
 
-        try:
-            s3.head_object(Bucket=bucket, Key=output_filename)
+        if not pipeline["settings"]["FUNCTION_FORCE_COMPUTE"]:
 
-            # Object exists, so we don't have to recompute
-            # TODO: the "force_compute" parameter will never work as is written in aws.py
-            if not pipeline.get("force_compute", False):
-                cached = True
+            # get configured s3 bucket to check for cache
+            bucket = pipeline["settings"]["S3_BUCKET_NAME"]
 
-        except botocore.exceptions.ClientError:
-            pass
+            # We can return if there is a valid cached output to save compute time.
+            try:
+                s3.head_object(Bucket=bucket, Key=pipeline["output"]["filename"])
+                return None
 
-        if cached:
-            return
+            # throws ClientError if no file is found
+            except botocore.exceptions.ClientError:
+                pass
+
+        # return pipeline definition
+        return pipeline
+
+    elif trigger == "APIGateway":
+        print("Triggered from API Gateway")
+
+        pipeline = default_pipeline()
+        pipeline["url"] = event["queryStringParameters"]
+        pipeline["params"] = urllib.parse_qs(urllib.urlparse(pipeline["url"]).query)
```

Comment:
We actually don't need to do any URL parsing, because `event['queryStringParameters']` is a dict with all of the params already parsed.